Content.Server/GameObjects/Components/Temperature/TemperatureComponent.cs

```diff
@@ -1,61 +1,107 @@
+        private float _currentTemperature;
+        private float _specificHeat;
+
+        private float _secondsSinceLastDamageUpdate = 0;
 
         public override void ExposeData(ObjectSerializer serializer)
         {
             base.ExposeData(serializer);
 
-            serializer.DataField(ref _fireDamageThreshold, "firedamagethreshold", 0);
-            serializer.DataField(ref _fireDamageCoefficient, "firedamagecoefficient", 1);
+            serializer.DataField(ref _fireDamageThreshold, "fireDamageThreshold", 0);
+            serializer.DataField(ref _coldDamageThreshold, "coldDamageThreshold", 0);
+            serializer.DataField(ref _tempDamageCoefficient, "tempDamageCoefficient", 1);
+            serializer.DataField(ref _currentTemperature, "currentTemperature", Atmospherics.T20C);
+            serializer.DataField(ref _specificHeat, "specificHeat", Atmospherics.MinimumHeatCapacity);
         }
 
-        /// <inheritdoc />
-        public void OnUpdate(float frameTime)
+        public void Update(float frameTime)
         {
-            var fireDamage =
-                (int) Math.Floor(Math.Max(0, CurrentTemperature - _fireDamageThreshold) / _fireDamageCoefficient);
+            var tempDamage = 0;
+            DamageType? damageType = null;
+            if (CurrentTemperature >= _fireDamageThreshold)
+            {
+                tempDamage = (int) Math.Floor((CurrentTemperature - _fireDamageThreshold) * _tempDamageCoefficient);
+                damageType = DamageType.Heat;
+            }
+            else if (CurrentTemperature <= _coldDamageThreshold)
+            {
+                tempDamage = (int) Math.Floor((_coldDamageThreshold - CurrentTemperature) * _tempDamageCoefficient);
+                damageType = DamageType.Cold;
+            }
+
+            if (damageType.HasValue)
+            {
+                _secondsSinceLastDamageUpdate += frameTime;
+                HandleTemperatureDamage(damageType.Value, tempDamage);
+                return;
+            }
+        }
 
-            _secondsSinceLastDamageUpdate += frameTime;
+        // Set new temperature
+        public void SetTemperature(float newTemperature)
+        {
+            // creadth: if we won't have anything useful here
+            // we might probably replace it with public setter
+            // on the actual property instead of separate method
+            _currentTemperature = newTemperature;
+        }
 
-            Owner.TryGetComponent(out IDamageableComponent component);
+        private void HandleTemperatureDamage(DamageType type, int amount)
+        {
+            if (!Owner.TryGetComponent(out IDamageableComponent component)) return;
 
             while (_secondsSinceLastDamageUpdate >= 1)
             {
-                component?.ChangeDamage(DamageType.Heat, fireDamage, false, null);
+                component?.ChangeDamage(type, amount, false);
                 _secondsSinceLastDamageUpdate -= 1;
             }
```

Comment:
Does this really have to be a while loop? It feels like there should be a better way to handle this.